src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ReflectionVirtualInvokeMapNode.cs

```diff
@@ -0,0 +1,224 @@
+                // Virtual Method uses a normal slot 
+                // TypeKey + NameAndSig metadata offset into the native layout metadata + (NumberOfStepsUpParentHierarchyToType << 1) + slot
+                // OR
+                // Generic Virtual Method 
+                // TypeKey + NameAndSig metadata offset into the native layout metadata + (NumberOfStepsUpParentHierarchyToType << 1 + 1)
+
+                MethodDesc declaringMethodForSlot = MetadataVirtualMethodAlgorithm.FindSlotDefiningMethodForVirtualMethod(method.GetTypicalMethodDefinition());
+                uint parentHierarchyDistance = 0;
+                TypeDesc typeOfDeclaringMethodForSlot = declaringMethodForSlot.OwningType.GetTypeDefinition();
+                TypeDesc currentType = method.OwningType.GetTypeDefinition();
+
+                while (typeOfDeclaringMethodForSlot != currentType)
+                {
+                    parentHierarchyDistance++;
+                    currentType = currentType.BaseType.GetTypeDefinition();
+                }
+
+                Vertex vertex = null;
+
+                ISymbolNode containingTypeKeyNode = factory.NecessaryTypeSymbol(containingTypeKey);
+                NativeLayoutMethodNameAndSignatureVertexNode nameAndSig = factory.NativeLayout.MethodNameAndSignatureVertex(method.GetTypicalMethodDefinition());
+                NativeLayoutPlacedSignatureVertexNode placedNameAndSig = factory.NativeLayout.PlacedSignatureVertex(nameAndSig);
+
+                if (method.HasInstantiation)
+                {
+                    vertex = writer.GetTuple(
+                        writer.GetUnsignedConstant(_externalReferences.GetIndex(containingTypeKeyNode)),
+                        writer.GetUnsignedConstant((uint)placedNameAndSig.SavedVertex.VertexOffset),
+                        writer.GetUnsignedConstant((parentHierarchyDistance << 1) + GenericVirtualMethod));
+                }
+                else
+                {
+                    // Get the declaring method for slot on the instantiated declaring type
+                    if (method.OwningType.HasInstantiation)
+                    {
+                        TypeDesc containingTypeOfDeclaringMethodForSlot = method.OwningType;
+                        for (int i = 0; i < parentHierarchyDistance; i++)
+                            containingTypeOfDeclaringMethodForSlot = containingTypeOfDeclaringMethodForSlot.BaseType;
+                        declaringMethodForSlot = containingTypeOfDeclaringMethodForSlot.GetMethod(declaringMethodForSlot.Name, declaringMethodForSlot.Signature);
+                    }
+
+                    int slot = VirtualMethodSlotHelper.GetVirtualMethodSlot(factory, declaringMethodForSlot, factory.Target.Abi == TargetAbi.ProjectN);
+
+                    if (slot == -1)
+                    {
+                        // This method doesn't have a slot. (At this time, this is only done for the Object.Finalize method)
```

Comment:
No, because the Finalize() method is virtual (so it might show up in this loop), but it's not in the vtable, so we just skip it.

---
In reply to: [108006416](https://github.com/dotnet/corert/pull/3100#discussion_r108006416) [](ancestors = 108006416)